### PR TITLE
Add new Retry Modes

### PR DESCRIPTION
**Pull request by mullermp**
Adds standard and adaptive retry modes, moves the current retry mode to legacy.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: gems/aws-sdk-core/lib/aws-sdk-core/plugins/retry_errors.rb [hunk 1]
```diff
@@ -1,177 +1,224 @@
...
 
-        CHECKSUM_ERRORS = Set.new([
-          'CRC32CheckFailed', # dynamodb
-        ])
+      option(
+        :retry_mode,
+        default: 'standard',
+        doc_type: String,
+        docstring: <<-DOCS) do |cfg|
+Specifies which retry algorithm to use. Defaults to `legacy`
```
Comment (mullermp) [inline gems/aws-sdk-core/lib/aws-sdk-core/plugins/retry_errors.rb]: Let's beef this up

File: gems/aws-sdk-core/lib/aws-sdk-core/plugins/retries/retry_quota.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+      # @api private
+      # Used in 'standard' and 'adaptive' retry modes.
+      class RetryQuota
+        INITIAL_RETRY_TOKENS = 500
+        RETRY_COST = 5
+        NO_RETRY_INCREMENT = 1
+        TIMEOUT_RETRY_COST = 10
+
+        def initialize
```
Comment (mullermp) [inline gems/aws-sdk-core/lib/aws-sdk-core/plugins/retries/retry_quota.rb]: Let's configure max capacity, and set available capacity to it. Default to initial

File: gems/aws-sdk-core/lib/aws-sdk-core/plugins/retry_errors.rb [hunk 3]
```diff
@@ -1,177 +1,224 @@
...
-          'PriorRequestNotComplete',                # route53
-        ])
+      # END LEGACY OPTIONS
 
-        CHECKSUM_ERRORS = Set.new([
-          'CRC32CheckFailed', # dynamodb
-        ])
+      option(
+        :retry_mode,
+        default: 'standard',
```
Comment (mullermp) [inline gems/aws-sdk-core/lib/aws-sdk-core/plugins/retry_errors.rb]: legacy

File: gems/aws-sdk-core/spec/aws/plugins/retries/client_rate_limiter_spec.rb [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        client_rate_limiter.send(:token_bucket_update_rate, 10)
+        client_rate_limiter.instance_variable_set(:@max_capacity, 100)
+
+        expect(mutex).to receive(:sleep).with(10) # 100 / fill_rate = 10
+        allow(Util).to receive(:monotonic_seconds).and_return(0, 10)
+
+        client_rate_limiter.token_bucket_acquire(100, true)
+      end
+
+      it 'raises a CapacityNotAvailableError when non block mode fails to acquire a token' do
```
Comment (mullermp) [inline gems/aws-sdk-core/spec/aws/plugins/retries/client_rate_limiter_spec.rb]: Change non block mode

File: gems/aws-sdk-core/spec/aws/plugins/retries/client_rate_limiter_spec.rb [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+
+        # Request a new token every 100ms (10 TPS) for 2 seconds.
+        expect(mutex).not_to receive(:sleep)
+        (0..20).each do |t|
+          allow(Util).to receive(:monotonic_seconds).and_return(1 + 0.1 * t)
+          client_rate_limiter.token_bucket_acquire(1)
+        end
+      end
+
+      it 'wont set rate below the min' do
```
Comment (mullermp) [inline gems/aws-sdk-core/spec/aws/plugins/retries/client_rate_limiter_spec.rb]: min fill rate

File: gems/aws-sdk-core/spec/aws/plugins/retries/client_rate_limiter_spec.rb [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        expect(mutex).not_to receive(:sleep)
+        client_rate_limiter.token_bucket_acquire(1)
+      end
+
+      it 'enables the token bucket on throttling errors' do
+        client_rate_limiter.update_sending_rate(true)
+        expect(client_rate_limiter.instance_variable_get(:@enabled)).to be(true)
+      end
+
+      describe 'RetryErrors::ClientRateLimiter ThreadSafety' do
```
Comment (mullermp) [inline gems/aws-sdk-core/spec/aws/plugins/retries/client_rate_limiter_spec.rb]: context 'thread safety'

File: gems/aws-sdk-core/spec/aws/plugins/retries/error_inspector_spec.rb [hunk 7]
```diff
@@ -0,0 +1,136 @@
+require_relative '../../../spec_helper'
+require_relative '../../../support/retry_errors_helper'
+
+module Aws
+  module Plugins
+    describe Retries::ErrorInspector do
```
Comment (mullermp) [inline gems/aws-sdk-core/spec/aws/plugins/retries/error_inspector_spec.rb]: Missing tests for modeled exceptions and endpoint discovery

File: gems/aws-sdk-core/spec/aws/plugins/retries/handler_spec.rb [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+              response: { status_code: 500, error: service_error },
+              expect: { available_capacity: 490, retries: 2 }
+            } # failure
+          ]
+
+          run_retry(test_case_def)
+        end
+
+        it 'fails due to retry quota reached after a single retry' do
+          resp.context.config.retry_quota.instance_variable_set(:@available_capacity, 5)
```
Comment (mullermp) [inline gems/aws-sdk-core/spec/aws/plugins/retries/handler_spec.rb]: Can possibly remove resp.context

File: gems/aws-sdk-core/spec/aws/plugins/retries/handler_spec.rb [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+          run_retry success(5, 7.0)
+          run_retry success(6, 9.6)
+          run_retry success(7, 10.0)
+          run_retry success(8, 10.45)
+          run_retry success(9, 13.4)
+          run_retry success(10, 21.2)
+          run_retry success(11, 36.4)
+        end
+
+        it 'verifies cubic calculations with throttling' do
```
Comment (mullermp) [inline gems/aws-sdk-core/spec/aws/plugins/retries/handler_spec.rb]: Let's delete

File: gems/aws-sdk-core/spec/aws/plugins/retries/legacy_handler_spec.rb [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+      let(:handler) { RetryErrors::LegacyHandler.new }
+
+      before(:each) do
+        resp.context.config = config
+        operation.endpoint_discovery = {}
+        resp.context.operation = operation
+        resp.context.http_response.status_code = 400
+      end
+
+      def handle(send_handler = nil, &block)
```
Comment (mullermp) [inline gems/aws-sdk-core/spec/aws/plugins/retries/legacy_handler_spec.rb]: Can remove

File: gems/aws-sdk-core/spec/aws/plugins/retry_errors_spec.rb [hunk 11]
```diff
@@ -1,9 +1,9 @@
 require_relative '../../spec_helper'
+require_relative '../../support/retry_errors_helper'
```
Comment (mullermp) [inline gems/aws-sdk-core/spec/aws/plugins/retry_errors_spec.rb]: We can add tests here that validate given a config, the appropriate handler is loaded

File: gems/aws-sdk-core/spec/support/retry_errors_helper.rb [hunk 12]
```diff
@@ -0,0 +1,82 @@
+
+RetryErrorsSvc = ApiHelper.sample_service
+
+# Sets up the handler to run retry tests
+# by calling either the send_handler or passed block
+# handler and resp must be defined outside this helper
```
Comment (mullermp) [inline gems/aws-sdk-core/spec/support/retry_errors_helper.rb]: Requires a handler defined

File: gems/aws-sdk-core/spec/support/retry_errors_helper.rb [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+  response = test_case[:response]
+  resp.context.http_response.status_code = response[:status_code]
+  resp.error = response[:error]
+
+  if response[:timestamp]
+    allow(Aws::Util).to receive(:monotonic_seconds).and_return(response[:timestamp])
+  end
+end
+
+def success(timestamp, calculated_rate)
```
Comment (mullermp) [inline gems/aws-sdk-core/spec/support/retry_errors_helper.rb]: Refactor/delete these

File: gems/aws-sdk-core/spec/aws/plugins/retry_errors_spec.rb [hunk 14]
```diff
@@ -291,6 +291,54 @@ module Plugins
...
+
+        it 'verifies success and throttling behavior' do
+          client_rate_limiter = config.client_rate_limiter
+          client_rate_limiter.instance_variable_set(:@last_throttle_time, 0)
+          # Needs to be smaller than 't' in the iterations
+          client_rate_limiter.instance_variable_set(:@last_tx_rate_bucket, 0)
+          client_rate_limiter.instance_variable_set(:@last_max_rate, 0)
+
+
+          def success(timestamp, measured_tx_rate, fill_rate)
```
Comment (mullermp) [inline gems/aws-sdk-core/spec/aws/plugins/retry_errors_spec.rb]: We might as well add these to the helper again and use them in the other tests

Comment (alextwoods) [reply gems/aws-sdk-core/spec/aws/plugins/retry_errors_spec.rb]: They are actually testing different expectations (eg the cubic calculations for success test is expecting calculated rate and this "e2e" test is looking at fill rate and measured_tx_rate.  I think it makes more sense to leave these local to this test only.

File: gems/aws-sdk-core/lib/aws-sdk-core/plugins/retries/retry_quota.rb [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+        NO_RETRY_INCREMENT = 1
+        TIMEOUT_RETRY_COST = 10
+
+        def initialize(opts = {})
+          @mutex              = Mutex.new
+          @max_capacity       = opts.fetch(:max_capacity, INITIAL_RETRY_TOKENS)
+          @available_capacity = @max_capacity
+        end
+
+        # check if there is sufficient capacity to retry
```
Comment (alextwoods) [inline gems/aws-sdk-core/lib/aws-sdk-core/plugins/retries/retry_quota.rb]: Update documentation to reflect what this actually does

File: gems/aws-sdk-core/lib/aws-sdk-core/plugins/retries/clock_skew.rb [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+    module Retries
+
+      # @api private
+      class ClockSkew
+
+        CLOCK_SKEW_THRESHOLD = 5 * 60 # five minutes
+
+        def initialize
+          @mutex = Mutex.new
+          @system_clock_offset = 0
```
Comment (alextwoods) [inline gems/aws-sdk-core/lib/aws-sdk-core/plugins/retries/clock_skew.rb]: Unused: remove

File: gems/aws-sdk-core/lib/aws-sdk-core/plugins/retries/clock_skew.rb [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+
+        # Update the stored clock skew value for an endpoint
+        # from the server's time in the response
+        # @param context [Seahorse::Client::RequestContext]
+        def update_clock_skew(context)
+          endpoint = context.http_request.endpoint
+          now_utc = Time.now.utc
+          server_time = server_time(context.http_response)
+          if server_time && (now_utc - server_time).abs > CLOCK_SKEW_THRESHOLD
+            set_clock_correction(endpoint, server_time - now_utc)
```
Comment (alextwoods) [inline gems/aws-sdk-core/lib/aws-sdk-core/plugins/retries/clock_skew.rb]: Should we warn the first time this is detected (other SDKs, eg .net print a warning)

File: gems/aws-sdk-core/lib/aws-sdk-core/shared_config.rb [hunk 18]
```diff
@@ -149,10 +149,14 @@ def self.config_reader(*attrs)
...
       :csm_port,
       :endpoint_discovery_enabled,
+      :max_attempts,
       :region,
+      :retry_mode,
       :s3_use_arn_region,
       :s3_us_east_1_regional_endpoint,
-      :sts_regional_endpoints
+      :sts_regional_endpoints,
+      :adaptive_retry_wait_to_fill,
```
Comment (alextwoods) [inline gems/aws-sdk-core/lib/aws-sdk-core/shared_config.rb]: Alphabetize

File: gems/aws-sdk-core/lib/aws-sdk-core/plugins/retries/clock_skew.rb (hunk 17 above)
Comment (mullermp) [reply gems/aws-sdk-core/lib/aws-sdk-core/plugins/retries/clock_skew.rb]: Talked offline.. could get annoying.

**Pull request merged**